### PR TITLE
Support for Distinct clause

### LINKED ISSUE #18: Feature: DISTINCT clause
Support for DISTINCT clause
Example queries: 
1. `SELECT DISTINCT sid, sname FROM student`
2. `SELECT DISTINCT sid, sname from student, enroll WHERE sid = studentid`
3. `SELECT DISTINCT sid, sname, eid from student, enroll WHERE sid = studentid`

### PR DESCRIPTION
**Pull request by xinyee20**
## Changes in this PR

- [ ] Bug fix
- [x] New feature
- [ ] Breaking change of an existing feature
- [ ] Refactor/Performance enhancements

## Overview
- Allow query to take in distinct keyword
- Add support for distinct clause (DistinctPlan and DistinctScan)
- Fixes #18

### CODE REVIEW

**Comment by xinyee20** [comment]
Test queries
`select did from dept, course`
Projection result:
```
 did
----
 10
 10
 10
 10
 10
 10
 20
 20
 20
 20
 20
 20
 30
 30
 30
 30
 30
 30
```

File: src/simpledb/query/DistinctScan.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        }
+        if (sameRecord)
+          hasmore = s.next();
+        else {
+          for (int i = 0; i < fieldlist.size(); i++) {
+            String field = fieldlist.get(i);
+            previousRecord.set(i, s.getVal(field));
+          }
+          return true;
+        }
```
Comment (don-tay) [inline src/simpledb/query/DistinctScan.java:63]: Alternatively
```suggestion
        for (int i = 0; i < fieldlist.size(); i++) {
          String field = fieldlist.get(i);
          Constant value = s.getVal(field);
          Constant prevRecVal = previousRecord.get(i);
          if (value.compareTo(prevRecVal) != 0) {
            for (int j = 0; j < fieldlist.size(); j++) {
              String field2 = fieldlist.get(j);
              previousRecord.set(j, s.getVal(field2));
            }
            return true;
          }
        }
        hasmore = s.next();
```

**Review by don-tay — APPROVED** [review]
lgtm, great work!

File: src/simpledb/query/DistinctScan.java [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        return true;
+      } else {
+        boolean sameRecord = true;
+        for (int i = 0; i < fieldlist.size(); i++) {
+          String field = fieldlist.get(i);
+          Constant value = s.getVal(field);
+          Constant prevRecVal = previousRecord.get(i);
+          if (value.compareTo(prevRecVal) != 0)
+            sameRecord = false;
```
[the remark this answers is not in the document]
Comment (xinyee20) [reply src/simpledb/query/DistinctScan.java]: Adjusted to new alternative in c42ee6e

File: src/simpledb/query/DistinctScan.java (hunk 1 above)
Comment (xinyee20) [reply src/simpledb/query/DistinctScan.java]: Updated in c42ee6e

**Pull request merged**